core/src/main/java/com/linecorp/armeria/server/annotation/JavaBeanRequestConverterFunction.java

```diff
@@ -0,0 +1,181 @@
+package com.linecorp.armeria.server.annotation;
+
+import static com.linecorp.armeria.server.AnnotatedHttpServiceParamUtil.convertParameter;
+import static com.linecorp.armeria.server.AnnotatedHttpServiceParamUtil.httpParametersOf;
+import static org.reflections.ReflectionUtils.getAllFields;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.AggregatedHttpMessage;
+import com.linecorp.armeria.common.HttpParameters;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.server.AnnotatedHttpServiceParamUtil.EnumConverter;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.util.AsciiString;
+
+public class JavaBeanRequestConverterFunction implements RequestConverterFunction {
+    private final Map<Class<?>, BeanParamInfo> beanParamsMap = new HashMap<>();
+
+    @Override
+    public Object convertRequest(final ServiceRequestContext ctx,
+                                 final AggregatedHttpMessage request,
+                                 final Class<?> expectedResultType) throws Exception {
+        if (!shouldHandle(request)) {
+            return RequestConverterFunction.fallthrough();
+        }
+
+        final BeanParamInfo beanParamInfo =
+                beanParamsMap.computeIfAbsent(expectedResultType,
+                                              JavaBeanRequestConverterFunction::buildBeanParamInfo);
+
+        final Object requestBean = beanParamInfo.createInstance();
+
+        HttpParameters httpParams = null;
+        for (final FieldParamInfo fieldParamInfo : beanParamInfo.fieldParamList) {
+            if (fieldParamInfo.paramName != null) {
+                if (httpParams == null) {
+                    httpParams = httpParametersOf(ctx, request.headers(), request);
+                }
+
+                String paramValueStr = ctx.pathParam(fieldParamInfo.paramName);
+                if (paramValueStr == null) {
+                    paramValueStr = httpParams.get(fieldParamInfo.paramName);
+                }
+
+                fieldParamInfo.setValue(requestBean, paramValueStr);
+            } else if (fieldParamInfo.headerName != null) {
+                final String headerValueStr = request.headers().get(AsciiString.of(fieldParamInfo.headerName));
+
+                fieldParamInfo.setValue(requestBean, headerValueStr);
+            }
+        }
+
+        return requestBean;
+    }
+
+    private static boolean shouldHandle(final AggregatedHttpMessage request) {
+        final MediaType contentType = request.headers().contentType();
+        if (contentType != null && (contentType.is(MediaType.JSON) ||
+                                    contentType.subtype().endsWith("+json"))) {
+            // request in JSON format, should be handled by {@link JacksonRequestConverterFunction}
```

Comment:
nit: You don't need to use `{@link}` here because it's not a part of Javadoc. :-)

Reply:
fixed.